cty/function/function.go

```diff
@@ -244,19 +244,21 @@ func (f Function) Call(args []cty.Value) (val cty.Value, err error) {
 			return cty.UnknownVal(expectedType), nil
 		}
 
-		if val.IsMarked() && !spec.AllowMarked {
-			unwrappedVal, marks := val.Unmark()
-			// In order to avoid additional overhead on applications that
-			// are not using marked values, we copy the given args only
-			// if we encounter a marked value we need to unmark. However,
-			// as a consequence we end up doing redundant copying if multiple
-			// marked values need to be unwrapped. That seems okay because
-			// argument lists are generally small.
-			newArgs := make([]cty.Value, len(args))
-			copy(newArgs, args)
-			newArgs[i] = unwrappedVal
-			resultMarks = append(resultMarks, marks)
-			args = newArgs
+		if !spec.AllowMarked {
+			unwrappedVal, marks := val.UnmarkDeep()
```

Comment:
I was going to suggest using `val.ContainsMarked` rather than `val.Unmark` here on the grounds that the old code here was trying to avoid reconstructing data structures unless it actually encountered something marked. However, having reviewed both functions it seems like `ContainsMarked` is currently potentially rather expensive too -- at the very least, it will allocate an array deep enough to represent the deepest `cty.Path` within the data structure -- so maybe it's not a big deal which to use.

Given that though, I wonder if the complexity of only conditionally copying the `args` array (as described in the comment below) is warranted anymore. Maybe we should now just `args` in all cases and just conditionally assign over zero or more of its elements, to make this overall easier to follow.

What do you think? I'm not _greatly_ bothered by this, but I didn't really like the complexity of this conditional copying in the first place and had only really done it because `val.IsMarked()` is a relatively cheap operation that avoided doing a bunch of work in the common case.

Reply:
My thinking here was that `spec.AllowMarked` is almost always going to be false. Of ~80 stdlib functions, only 9 currently support marked parameters. When `UnmarkDeep` is potentially so expensive, I felt the complexity trade-off was reasonable.